modulemd/v2/modulemd-module-index.c

```diff
@@ -0,0 +1,536 @@
+    {
+      module = (ModulemdModule *)value;
+
+      if (!dump_defaults (module, emitter, error))
+        return FALSE;
+
+      if (!dump_translations (module, emitter, error))
+        return FALSE;
+
+      if (!dump_streams (module, emitter, error))
+        return FALSE;
+    }
+
+  if (!mmd_emitter_end_stream (emitter, error))
+    return FALSE;
+
+  return TRUE;
+}
+
+
+gboolean
+modulemd_module_index_update_from_file (ModulemdModuleIndex *self,
+                                        const gchar *yaml_file,
+                                        GPtrArray **failures,
+                                        GError **error)
+{
+  g_return_val_if_fail (MODULEMD_IS_MODULE_INDEX (self), FALSE);
+
+  g_autoptr (FILE) yaml_stream = NULL;
+
+  yaml_stream = g_fopen (yaml_file, "rb");
+
+  return modulemd_module_index_update_from_stream (
+    self, yaml_stream, failures, error);
+}
+
+
+gboolean
+modulemd_module_index_update_from_string (ModulemdModuleIndex *self,
+                                          const gchar *yaml_string,
+                                          GPtrArray **failures,
+                                          GError **error)
+{
+  g_return_val_if_fail (MODULEMD_IS_MODULE_INDEX (self), FALSE);
+
+  MMD_INIT_YAML_PARSER (parser);
+
+  yaml_parser_set_input_string (
+    &parser, (const unsigned char *)yaml_string, strlen (yaml_string));
+
+  return modulemd_module_index_update_from_parser (
+    self, &parser, failures, error);
+}
+
+
+gboolean
+modulemd_module_index_update_from_stream (ModulemdModuleIndex *self,
+                                          FILE *yaml_stream,
+                                          GPtrArray **failures,
+                                          GError **error)
+{
+  g_return_val_if_fail (MODULEMD_IS_MODULE_INDEX (self), FALSE);
+
+  MMD_INIT_YAML_PARSER (parser);
+
+  yaml_parser_set_input_file (&parser, yaml_stream);
+
+  return modulemd_module_index_update_from_parser (
+    self, &parser, failures, error);
+}
+
+
+const gchar *
+modulemd_module_index_dump_to_string (ModulemdModuleIndex *self,
+                                      GError **error)
+{
+  g_return_val_if_fail (MODULEMD_IS_MODULE_INDEX (self), NULL);
+
+  MMD_INIT_YAML_EMITTER (emitter);
+  MMD_INIT_YAML_STRING (&emitter, yaml_string);
+
+  if (!modulemd_module_index_dump_to_emitter (self, &emitter, error))
+    return NULL;
+
+  return g_strdup (yaml_string->str);
+}
+
+
+gboolean
+modulemd_module_index_dump_to_stream (ModulemdModuleIndex *self,
+                                      FILE *yaml_stream,
+                                      GError **error)
+{
+  g_return_val_if_fail (MODULEMD_IS_MODULE_INDEX (self), FALSE);
+
+  MMD_INIT_YAML_EMITTER (emitter);
+  yaml_emitter_set_output_file (&emitter, yaml_stream);
+
+  return modulemd_module_index_dump_to_emitter (self, &emitter, error);
+}
+
+
+gboolean
+modulemd_module_index_validate (ModulemdModuleIndex *self, GError **error)
```

Comment:
Let's drop the `validate()` function for now (or at least make it non-public). I'm not sure we want this to be a simple boolean or if we want to report any and all failures on contained objects in some other manner.

Rather than design it right now, let's just avoid making it public API.